cpp/benchmarks/string/contains.cpp

```diff
@@ -19,48 +19,99 @@
+}
+
 enum contains_type { contains, count, findall };
 
+// longer pattern lengths demand more working memory per string
+std::string patterns[] = {"^\\d+ [a-z]+", "[A-Z ]+\\d+ +\\d+[A-Z]+\\d+$"};
+
 static void BM_contains(benchmark::State& state, contains_type ct)
 {
-  cudf::size_type const n_rows{(cudf::size_type)state.range(0)};
-  auto const table = create_random_table({cudf::type_id::STRING}, row_count{n_rows});
-  cudf::strings_column_view input(table->view().column(0));
+  auto const n_rows        = static_cast<cudf::size_type>(state.range(0));
+  auto const pattern_index = static_cast<int32_t>(state.range(1));
+  auto const hit_rate      = static_cast<int32_t>(state.range(2));
+
+  auto col   = build_input_column(n_rows, hit_rate);
+  auto input = cudf::strings_column_view(col->view());
+
+  auto pattern = patterns[pattern_index];
 
   for (auto _ : state) {
     cuda_event_timer raii(state, true, rmm::cuda_stream_default);
-    // contains_re(), matches_re(), and count_re() all have similar functions
-    // with count_re() being the most regex intensive
     switch (ct) {
       case contains_type::contains:  // contains_re and matches_re use the same main logic
-        cudf::strings::contains_re(input, "\\d+");
+        cudf::strings::contains_re(input, pattern);
         break;
-      case contains_type::count:  // counts occurrences of pattern
-        cudf::strings::count_re(input, "\\d+");
+      case contains_type::count:  // counts occurrences of matches
+        cudf::strings::count_re(input, pattern);
         break;
-      case contains_type::findall:  // returns occurrences of matches
-        cudf::strings::findall(input, "\\d+");
+      case contains_type::findall:  // returns occurrences of all matches
+        cudf::strings::findall(input, pattern);
         break;
     }
   }
 
   state.SetBytesProcessed(state.iterations() * input.chars_size());
 }
 
-#define STRINGS_BENCHMARK_DEFINE(name, b)                          \
-  BENCHMARK_DEFINE_F(StringContains, name)                         \
-  (::benchmark::State & st) { BM_contains(st, contains_type::b); } \
-  BENCHMARK_REGISTER_F(StringContains, name)                       \
-    ->RangeMultiplier(8)                                           \
-    ->Ranges({{1 << 12, 1 << 24}})                                 \
-    ->UseManualTime()                                              \
+#define STRINGS_BENCHMARK_DEFINE(name, b)                                         \
+  BENCHMARK_DEFINE_F(StringContains, name)                                        \
+  (::benchmark::State & st) { BM_contains(st, contains_type::b); }                \
+  BENCHMARK_REGISTER_F(StringContains, name)                                      \
+    ->ArgsProduct({{4096, 32768, 262144, 2097152, 16777216}, /* row count */      \
+                   {0, 1},                                   /* patterns index */ \
+                   {1, 5, 10, 25, 70, 100}})                 /* hit rate */       \
```

Comment:
This algorithm is much better. I would like to see `hit_rate` be a float between 0 and 1. That would align better with existing patterns like `set_null_frequency`, which accepts a float between 0 and 1. Alternatively, I suppose you could document that it is an integer percentage between 0 and 100.

Reply:
Also maybe "hit rate" should be named "match frequency"? Just trying to be consistent in terms, between hit and match.

Reply:
I'd rather not use float values unless necessary. The inaccuracies can cause issues. For example, the `set_null_frequency` is not working perfectly because I got a single null row for a column of 16M entries even though I set the null frequency to 0.0. I'm guessing there is some floating point rounding error in the generate logic.